### ISSUE TITLE
Use peerDependency for pug

### ISSUE THREAD

**Issue by TimothyGu** [gulp-community/gulp-pug#135]
This way users can prevent downloading two copies of Pug (one bundled with gulp-pug, the other their own). This is especially important as Pug currently hasn't released a stable version, so bug fixes are fast in coming.

**Comment by shinnn** [comment]
Why not use [`opts.pug`](https://github.com/jamen/gulp-pug#pugopts)?

> This way users can prevent downloading two copies of Pug (one bundled with gulp-pug, the other their own).

In this case, npm already de-duplicates those two `pug` packages well.

# 

As one of the [gulp-eslint](https://github.com/adametry/gulp-eslint) maintainers I've also received a lot of requests to switch `eslint` from `dependencies` to `peerDependencies`. However, [we finally decided not](https://github.com/adametry/gulp-eslint/issues/24#issuecomment-71477071). `peerDependencies` introduces considerable confusion to users rather than solves a few problems.

For reference, [one of npm authors considers `peerDependencies` basically broken](https://github.com/npm/npm/issues/5080#issuecomment-40545599).

**Comment by jamen** [comment]
The [original `gulp-pug`](https://github.com/jamen-archive/gulp-pug) I made used `peerDependencies`.  After a while, I personally found it annoying, because npm3 doesn't download the package automatically.  It was an extra step that confused some users (especially those switching from `gulp-jade` to `gulp-pug`).  So I ended up removing it.

> This is especially important as Pug currently hasn't released a stable version

This also is bothersome because that means I'll have to use an asterisk for quantifying the version, which puts me in the blind for things like breaking changes:

``` javascript
"peerDependencies": {
  "pug": "*"
}
```

In other issues, I've commented that I will update the version in the `dependencies` once there is a proper non-deprecated/non-pre release:

> I'm not updating the dependencies until they get an actual (non-prereleased) version out, because semver doesn't capture any of the prereleased versions unless you are specific about it (like we are now), and the only non-prereleased versions are marked as deprecated. So, I am going to wait.

@shinnn has some good points on it too.

Once there is a stable `pug` release, I think either system would work fine.  I just personally feel the current system is less confusing, and also compatible with the original `gulp-jade`.

**Comment by jamen** [comment]
For reference, opened #136.

**Comment by zenekron** [comment]
### Versioning
All npm packages are supposed to use [semantic versioning](https://docs.npmjs.com/misc/semver) [[2]](http://semver.org/)

**Comment by zenekron** [comment]
**TL;DR:** You should add `gulp` to the `peerDependencies` and widen `pug`'s versions range (`^2.x`).

Directly from the official docs on [peerDependencies](https://docs.npmjs.com/files/package.json#peerdependencies):

> In some cases, you want to express the compatibility of your package with a host tool or library, while not necessarily doing a require of this host. This is usually referred to as a plugin. Notably, your module may be exposing a specific interface, expected and specified by the host documentation.

And it's true that `gulp-pug` has this kind of relationship.. **but with `gulp`, not with `pug`**!  
So it is my opinion that a _peer dependency_ should be added, and that would be in fact `gulp`.

Also I agree with @TimothyGu when he says that we should give users the possibility to "choose" whatever version of pug they want to use (while preventing errors due to breaking changes).

This can be done if `pug` follows the [semantic versioning](http://semver.org/) rules (and it **should**).
In fact semver states that patches and minor updates should be backwards compatible, so my proposal would be to target a wider range of pug versions by using `^2.x` as the version selector.

By doing the users are given the **possibility**, but not the _duty_ (as would happen if pug was a peer dependency), to choose which version of pug to use while ensuring its compatibility.

PS: I actually don't know if the pug devs are introducing breaking api changes across alpha versions (and they shouldn't); if so giving the users the freedom to choice the version might be unapplicable.

**Comment by jamen** [comment]
@zenekron Part of the problem I was trying to explain is that you cannot capture preleased versions with those selectors.  For example, `^2.0.0` in `peerDependencies` would be unmet for any of the v2 prereleases, and then the only versions that can be captured are deprecated.  

That would make it so either no versions are captured, or only deprecated versions are captured.  I would be forced to use `*` to let the user have a free choice with the prereleases. (Which removes any benefits of semver)

As far as letting users have a choice in what version they want to use, there is `opts.pug` as described.

**Comment by shinnn** [comment]
> You should add `gulp` to the `peerDependencies`

No. gulp plugins sometimes can be used outside gulp. For example [duo-gulp](https://github.com/duojs/gulp) converts a gulp plugin to a duo plugin, and we don't need to install `gulp` while using `gulp-pug` via `duo-gulp`.

Adding `gulp` to `peerDependencies` won't benefits users but just restrict its availability.

**Comment by zenekron** [comment]
@jamen I personally don't consider having an `opts.pug` good, and if the problem is just the version selector then `>=2.0.0-alpha <3` would do the trick..

@shinnn I wasn't aware of that use case for gulp plugins so thanks for pointing out!

**Comment by jamen** [comment]
@zenekron Sure, I can use that in the `dependencies`, but I still think `peerDependencies` confuses people (I had a few issues opened on the last repo related to it).  Especially people who are migrating from `gulp-jade`.

**Comment by zenekron** [comment]
@jamen In fact I'm strongly against having `pug` in the `peerDependencies` 😅

**Comment by TimothyGu** [comment]
Okay, those are some valid points against `peerDependencies`, so I'll drop this issue, though that's probably not the way I would have gone.

But just to answer some questions about Pug versioning, yes, we do follow semver. We don't intentionally break compat between alpha/beta versions (Pug is now at beta), but when there is an API bug/inconsistency you have to do that. Furthermore, breaking compat between alpha versions is permitted per [semver #9](http://semver.org/#spec-item-9).

Currently, the issue holding Pug back from a stable release is more because of a lack of updated documentation and website than it is code-wise. When we do release a stable version, I'll be sure to let you guys know.

**Comment by jamen** [comment]
Thank you @TimothyGu.
